nilearn/regions/parcellations.py

```diff
@@ -0,0 +1,429 @@
+"""Parcellation tools such as KMeans or Ward for fMRI images
+"""
+
+import numpy as np
+
+import sklearn
+from sklearn.base import clone
+from sklearn.feature_extraction import image
+from sklearn.externals.joblib import Memory, delayed, Parallel
+
+from ..decomposition.base import BaseDecomposition, mask_and_reduce
+from ..input_data import NiftiLabelsMasker
+from .._utils.compat import _basestring
+from .._utils.niimg import _safe_get_data
+
+
+def _estimator_fit(data, estimator):
+    """ Estimator to fit on the reduced data
+
+    Parameters
+    ----------
+    data : numpy arrays
+        list of reduced numpy arrays
+
+    estimator : instance of estimator from sklearn
+        MiniBatchKMeans or AgglomerativeClustering
+
+    Returns
+    -------
+    labels_ : numpy.ndarray
+        labels_ estimated from estimator
+    """
+    estimator = clone(estimator)
+    estimator.fit(data.T)
+
+    return estimator.labels_
+
+
+def _check_parameters_transform(imgs, confounds):
+    """A helper function to check the parameters and prepare for processing
+    as a list.
+    """
+    if not isinstance(imgs, (list, tuple)) or \
+            isinstance(imgs, _basestring):
+        imgs = [imgs, ]
+
+    if confounds is None and isinstance(imgs, (list, tuple)):
+        confounds = [None] * len(imgs)
+
+    if confounds is not None:
+        if not isinstance(confounds, (list, tuple)) or \
+                isinstance(confounds, _basestring):
+            confounds = [confounds, ]
+
+    if len(confounds) != len(imgs):
+        raise ValueError("Number of confounds given does not match with "
+                         "the given number of images.")
+    return imgs, confounds
+
+
+def _labels_masker_extraction(img, masker, confound):
+    """ Helper function for parallelizing NiftiLabelsMasker extractor
+    on list of Nifti images.
+
+    Parameters
+    ----------
+    img : 4D Nifti image like object
+        Image to process.
+
+    masker : instance of NiftiLabelsMasker
+        Used for extracting signals with fit_transform
+
+    confound : csv file or numpy array
+        Confound used for signal cleaning while extraction.
+        Passed to signal.clean
+
+    Returns
+    -------
+    signals : numpy array
+        Signals extracted on given img
+    """
+    signals = masker.fit_transform(img, confounds=confound)
```

Comment:
I think that we should clone the masker in this function. The reasoning is to avoid a different behavior when run in parallel and in series